### PR TITLE
Fix error in M4T feature extractor

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

Really small PR that fixes an error when calling the SeamlessM4TFeatureExtractor without attention mask. I simply added a check to verify if the attention mask exists before operating on it. I've also modified the test suite to test this in the future.

cc @amyeroberts or @ArthurZucker ! WDYT ?

### CODE REVIEW

File: src/transformers/models/seamless_m4t/feature_extraction_seamless_m4t.py [hunk 1]
```diff
@@ -271,14 +271,16 @@ def __call__(
         remainder = num_frames % self.stride
         if remainder != 0:
             input_features = input_features[:, :num_frames, :]
-            attention_mask = attention_mask[:, :num_frames]
+            if return_attention_mask:
+                attention_mask = attention_mask[:, :num_frames]
```
Comment (amyeroberts) [inline src/transformers/models/seamless_m4t/feature_extraction_seamless_m4t.py:275]: How expensive is having `return_attention_mask=True` in the `pad` call? The logic here could all be tidied up by having it as `True` on L261 and then only keep it in `padded_inputs` on 264 if `return_attention_mask=True`

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing! 

Fix looks fine in principle but there's still some attention_mask logic which has surprising behaviour (key still in returned outputs) and tests shouldn't be changed

File: tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py [hunk 2]
```diff
@@ -62,7 +62,7 @@ def __init__(
         feature_size=10,
         padding_value=0.0,
         sampling_rate=4_000,
-        return_attention_mask=True,
+        return_attention_mask=False,
```
[the remark this answers is not in the document]
Comment (ylacombe) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py:65]: It's tested in [`SequenceFeatureExtractionTestMixin.test_attention_mask`](https://github.com/huggingface/transformers/blob/4a66c0d95219bbeb91bebd010d75a29a5e5f3f43/tests/test_sequence_feature_extraction_common.py#L389-L424), but I can still add some more tests if you want!

File: src/transformers/models/seamless_m4t/feature_extraction_seamless_m4t.py [hunk 3]
```diff
@@ -271,14 +271,16 @@ def __call__(
...
         )
 
-        indices = np.arange(0, num_frames)
-        attention_mask = attention_mask[:, indices % self.stride == 1]
+        if return_attention_mask:
+            indices = np.arange(0, num_frames)
+            attention_mask = attention_mask[:, indices % self.stride == 1]
 
         padded_inputs["input_features"] = input_features
         padded_inputs["attention_mask"] = attention_mask
```
[the remark this answers is not in the document]
Comment (ylacombe) [reply src/transformers/models/seamless_m4t/feature_extraction_seamless_m4t.py:286]: Nice catch

**Comment by ylacombe** [comment]
Hey @amyeroberts, thanks for the quick review and the relevant feedback, I've modified the logic a bit, WDYT ? There's also a pending comment on testing for which I'll wait your return!

File: tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py:65]: These tests only test the `.pad` call to the feature extractor. Not the `attention_mask` logic when calling the feature extractor directly - which this PR addresses

[the remark this answers is not in the document]
Comment (ylacombe) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py:65]: You're right, I've added a test

[the remark this answers is not in the document]
Comment (ylacombe) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py:65]: Let me know if that works for you @amyeroberts !

[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py:65]: Let's take a step back so I can understand the changes: 
* why is the default being changed here? 
* what is the default behaviour of the feature extractor (outside of tests)? Does it return the attention masks by default?

[the remark this answers is not in the document]
Comment (ylacombe) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py:65]: I've realized when copying the M4T feature extractor for another model there were errors when `return_attention_mask=False`, I've thus corrected the feature extractor code.

Now, with regards to tests, I realized it wasn't flagged by test so:

1. I've changed the default to `False`, because no current tests were actually using the attention mask.
2. The new test `test_call_attention_mask` checks that one can use the attention mask without errors.

I've done that despite the default behavior outside of tests being `return_attention_mask=True`.

Let me know if that responds to your questions. I can still change the default back to `True` and test without attention mask similarly to `test_call_attention_mask` if that's needed.

[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py:65]: OK, so in a perfect world we'd have full coverage of the behaviour of our objects, in this case with all different argument permutations and their expected output. As this isn't practical, it's important that the most common use cases are accounted for. All class methods are effectively black boxes: I do not know if the different input arguments interact or not. Therefore, it's easier and better when testing to use the defaults and interrogate the outputs based on changing a subset of the arguments.

[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py:65]: To be clear, a test should be added to check everything works when `return_attention_mask=False`, but it should be left as `True` here to make sure the default behaviour is compatible with what's happening in other tests

[the remark this answers is not in the document]
Comment (ylacombe) [reply tests/models/seamless_m4t/test_feature_extraction_seamless_m4t.py]: This makes total sense, thanks for the explanations here!  
I've modified the tests to go back to the initial situation but I've added two other tests to ensure better coverage.

I'll make sure to keep your explanation in mind for future PRs as well !

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for the quick fix and iterating on this!
